backend/app/balance_monitor.go

```diff
@@ -38,59 +66,303 @@ func (h *Handler) MonitorSystemBalanceAndHandleSettlement() {
+		return err
+	}
+
+	dailyUsageInTFT, err := internal.FromUSDMillicentToTFT(h.substrateClient, dailyUsageInUSDMillicent)
+	if err != nil {
+		logger.GetLogger().Error().Err(err).Send()
+		return err
+	}
+
+	totalPendingTFTAmount, err := h.db.CalculateTotalPendingTFTAmountPerUser(user.ID)
+	if err != nil {
+		logger.GetLogger().Error().Err(err).Send()
+		return err
+	}
+
+	userTFTBalance, err := internal.GetUserTFTBalance(h.substrateClient, user.Mnemonic)
+	if err != nil {
+		logger.GetLogger().Error().Err(err).Msgf("Failed to get user TFT balance for user %d", user.ID)
+		return err
+	}
+
+	// make sure no old payments will fund more than needed
+	if totalPendingTFTAmount < dailyUsageInTFT &&
+		userTFTBalance < dailyUsageInTFT &&
+		dailyUsageInTFT > 0 &&
+		user.CreditCardBalance+user.CreditedBalance-user.Debt < dailyUsageInUSDMillicent {
+		if err := h.db.CreateTransferRecord(&models.TransferRecord{
+			UserID:    user.ID,
+			Username:  user.Username,
+			TFTAmount: dailyUsageInTFT - userTFTBalance,
+			Operation: models.DepositOperation,
+		}); err != nil {
+			logger.GetLogger().Error().Err(err).Msgf("Failed to create transfer record for user %d", user.ID)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (h *Handler) calculateResourcesUsageInUSDApplyingDiscount(
+	ctx context.Context,
+	userID int,
+	userMnemonic string,
+	addedRentedNodes []types.Node,
+	addedSharedNodes []kubedeployer.Node,
+	configuredDiscount discount,
+) (uint64, error) {
+	userIdentity, err := substrate.NewIdentityFromSr25519Phrase(userMnemonic)
+	if err != nil {
+		return 0, errors.Wrapf(err, "Failed to create identity for user %d", userID)
+	}
+
+	calculator := calculator.NewCalculator(h.gridClient.SubstrateConn, userIdentity)
+
+	var totalResourcesCostMillicent uint64
+
+	rentedNodes, _, err := h.getRentedNodesForUser(ctx, userID, true)
+	if err != nil {
+		return 0, err
+	}
+	rentedNodes = append(rentedNodes, addedRentedNodes...)
+
+	// Calculate rented nodes
+	for _, node := range rentedNodes {
+		resourcesCost, err := calculator.CalculateCost(
+			node.TotalResources.CRU,
+			uint64(node.TotalResources.MRU),
+			uint64(node.TotalResources.HRU),
+			uint64(node.TotalResources.SRU),
+			len(node.PublicConfig.Ipv4) > 0,
+			len(node.CertificationType) > 0,
+		)
+		if err != nil {
+			return 0, err
+		}
+
+		// resources cost per month
+		pricingPolicy, err := h.substrateClient.GetPricingPolicy(defaultPricingPolicyID)
+		if err != nil {
+			return 0, err
+		}
+		dedicatedDiscountPercentage := float64(pricingPolicy.DedicatedNodesDiscount / 100)
+		totalResourcesCostMillicent += internal.FromUSDToUSDMillicent(resourcesCost * dedicatedDiscountPercentage)
+	}
+
+	sharedNodes, err := h.getUserNodes(userID)
+	if err != nil {
+		return 0, err
+	}
+	sharedNodes = append(sharedNodes, addedSharedNodes...)
+
+	// Calculate shared nodes
+	for _, node := range sharedNodes {
```

Comment:
Not all the cluster's nodes are on shared nodes. Can you add a check to see if the node is rented by the user to continue?